File: src/LTAdmin/Repositories/ExamRepository.cs
using System.Data;
using LTAdmin.Data;
using LTAdmin.Models.Dto;
using LTAdmin.Models.Entities;

namespace LTAdmin.Repositories;

/// <summary>Dépôt des examens : sessions, épreuves, notes d'examen.</summary>
public sealed class ExamRepository : RepositoryBase
{
    public ExamRepository(AccessDatabase database) : base(database)
    {
    }

    // ----- SESSION_EXAM -----

    public SessionExam? GetSession(int idSession)
        => QuerySingle(
            $"SELECT * FROM {Q(Tables.SessionExam)} WHERE {Q("ID_SESSION")} = ?",
            MapSession, P(idSession));

    public List<SessionExam> ListSessions(int? idAnnee = null)
        => idAnnee.HasValue
            ? QueryList($"SELECT * FROM {Q(Tables.SessionExam)} WHERE {Q("ID_ANNEE")} = ? ORDER BY {Q("DATE_DEBUT")}, {Q("LIBELLE")}", MapSession, P(idAnnee.Value))
            : QueryList($"SELECT * FROM {Q(Tables.SessionExam)} ORDER BY {Q("ID_ANNEE")} DESC, {Q("DATE_DEBUT")}", MapSession);

    public int InsertSession(SessionExam session)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.SessionExam)} ({Q("ID_ANNEE")}, {Q("LIBELLE")}, {Q("NATURE")}, {Q("DATE_DEBUT")}, {Q("DATE_FIN")}, {Q("CLOTUREE")}) VALUES (?, ?, ?, ?, ?, ?)",
            P(session.IdAnnee), P(session.Libelle), P(session.Nature),
            P(session.DateDebut), P(session.DateFin), P(session.Cloturee ?? false));

    public int UpdateSession(SessionExam session)
        => Execute(
            $"UPDATE {Q(Tables.SessionExam)} SET {Q("ID_ANNEE")} = ?, {Q("LIBELLE")} = ?, {Q("NATURE")} = ?, {Q("DATE_DEBUT")} = ?, {Q("DATE_FIN")} = ?, {Q("CLOTUREE")} = ? WHERE {Q("ID_SESSION")} = ?",
            P(session.IdAnnee), P(session.Libelle), P(session.Nature),
            P(session.DateDebut), P(session.DateFin), P(session.Cloturee ?? false), P(session.IdSession));

    public int SetSessionCloturee(int idSession, bool cloturee)
        => Execute(
            $"UPDATE {Q(Tables.SessionExam)} SET {Q("CLOTUREE")} = ? WHERE {Q("ID_SESSION")} = ?",
            P(cloturee), P(idSession));

    public int DeleteSession(int idSession)
        => Execute($"DELETE FROM {Q(Tables.SessionExam)} WHERE {Q("ID_SESSION")} = ?", P(idSession));

    // ----- EPREUVE -----

    public Epreuve? GetEpreuve(int idEpreuve)
        => QuerySingle(
            $"SELECT * FROM {Q(Tables.Epreuve)} WHERE {Q("ID_EPREUVE")} = ?",
            MapEpreuve, P(idEpreuve));

    public EpreuveDetail? GetEpreuveDetail(int idEpreuve)
        => QuerySingle(EpreuveDetailSql() + $" WHERE EP.{Q("ID_EPREUVE")} = ?", MapEpreuveDetail, P(idEpreuve));

    public List<EpreuveDetail> ListEpreuves(int? idSession = null, int? idClasse = null)
    {
        var conditions = new List<string>();
        var parameters = new List<System.Data.OleDb.OleDbParameter>();
        if (idSession.HasValue) { conditions.Add($"EP.{Q("ID_SESSION")} = ?"); parameters.Add(P(idSession.Value)); }
        if (idClasse.HasValue) { conditions.Add($"EP.{Q("ID_CLASSE")} = ?"); parameters.Add(P(idClasse.Value)); }
        var sql = EpreuveDetailSql();
        if (conditions.Count > 0) sql += " WHERE " + string.Join(" AND ", conditions);
        sql += $" ORDER BY EP.{Q("DATE_EPREUVE")}, EP.{Q("ID_EPREUVE")}";
        return QueryList(sql, MapEpreuveDetail, parameters.ToArray());
    }

    private static string EpreuveDetailSql()
        => $"SELECT EP.*, SE.{Q("LIBELLE")} AS SESSION_LIB, C.{Q("LIBELLE")} AS CLASSE_LIB, " +
           $"M.{Q("MATIERE")} AS MATIERE_LIB, S.{Q("NOM_SALLE")} AS SALLE_LIB " +
           $"FROM (((({Q(Tables.Epreuve)} AS EP " +
           $"INNER JOIN {Q(Tables.SessionExam)} AS SE ON EP.{Q("ID_SESSION")} = SE.{Q("ID_SESSION")}) " +
           $"INNER JOIN {Q(Tables.Classe)} AS C ON EP.{Q("ID_CLASSE")} = C.{Q("ID_CLASSE")}) " +
           $"INNER JOIN {Q(Tables.Matiere)} AS M ON EP.{Q("CODE_MATIERE")} = M.{Q("CODE_MATIERE")}) " +
           $"LEFT JOIN {Q(Tables.Salle)} AS S ON EP.{Q("ID_SALLE")} = S.{Q("ID_SALLE")})";

    public int InsertEpreuve(Epreuve epreuve)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.Epreuve)} ({Q("ID_SESSION")}, {Q("ID_CLASSE")}, {Q("CODE_MATIERE")}, {Q("DATE_EPREUVE")}, " +
            $"{Q("HEURE_DEBUT")}, {Q("DUREE_MN")}, {Q("COEFFICIENT")}, {Q("BAREME")}, {Q("ID_SALLE")}, {Q("SURVEILLANT")}) " +
            "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            P(epreuve.IdSession), P(epreuve.IdClasse), P(epreuve.CodeMatiere), P(epreuve.DateEpreuve),
            P(epreuve.HeureDebut), P(epreuve.DureeMn), P(epreuve.Coefficient), P(epreuve.Bareme),
            P(epreuve.IdSalle), P(epreuve.Surveillant));

    public int UpdateEpreuve(Epreuve epreuve)
        => Execute(
            $"UPDATE {Q(Tables.Epreuve)} SET {Q("ID_SESSION")} = ?, {Q("ID_CLASSE")} = ?, {Q("CODE_MATIERE")} = ?, {Q("DATE_EPREUVE")} = ?, " +
            $"{Q("HEURE_DEBUT")} = ?, {Q("DUREE_MN")} = ?, {Q("COEFFICIENT")} = ?, {Q("BAREME")} = ?, {Q("ID_SALLE")} = ?, {Q("SURVEILLANT")} = ? " +
            $"WHERE {Q("ID_EPREUVE")} = ?",
            P(epreuve.IdSession), P(epreuve.IdClasse), P(epreuve.CodeMatiere), P(epreuve.DateEpreuve),
            P(epreuve.HeureDebut), P(epreuve.DureeMn), P(epreuve.Coefficient), P(epreuve.Bareme),
            P(epreuve.IdSalle), P(epreuve.Surveillant), P(epreuve.IdEpreuve));

    public int DeleteEpreuve(int idEpreuve)
        => Execute($"DELETE FROM {Q(Tables.Epreuve)} WHERE {Q("ID_EPREUVE")} = ?", P(idEpreuve));

    // ----- NOTE_EXAMEN -----

    public NoteExamen? GetNoteExamen(int idEpreuve, int idInscription)
        => QuerySingle(
            $"SELECT * FROM {Q(Tables.NoteExamen)} WHERE {Q("ID_EPREUVE")} = ? AND {Q("ID_INSCRIPTION")} = ?",
            MapNoteExamen, P(idEpreuve), P(idInscription));

    public List<NoteSaisieRow> ListGrilleSaisie(int idEpreuve, int idClasse)
        => QueryList(
            $"SELECT N.{Q("ID_NOTE_EX")} AS ID_NOTE, I.{Q("ID_INSCRIPTION")}, E.{Q("MATRICULE")}, E.{Q("NOM")}, E.{Q("PRENOM")}, " +
            $"N.{Q("VALEUR_NOTE")}, N.{Q("ABSENT")} " +
            $"FROM (({Q(Tables.Inscription)} AS I " +
            $"INNER JOIN {Q(Tables.Etudiant)} AS E ON I.{Q("ID_ETUDIANT")} = E.{Q("ID_ETUDIANT")}) " +
            $"LEFT JOIN {Q(Tables.NoteExamen)} AS N ON N.{Q("ID_INSCRIPTION")} = I.{Q("ID_INSCRIPTION")} AND N.{Q("ID_EPREUVE")} = ?) " +
            $"WHERE I.{Q("ID_CLASSE")} = ? ORDER BY E.{Q("NOM")}, E.{Q("PRENOM")}",
            EvaluationRepository.MapNoteSaisieRow, P(idEpreuve), P(idClasse));

    public int CountNotesByEpreuve(int idEpreuve)
        => ScalarInt($"SELECT COUNT(*) FROM {Q(Tables.NoteExamen)} WHERE {Q("ID_EPREUVE")} = ?", P(idEpreuve));

    public int InsertNoteExamen(NoteExamen note)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.NoteExamen)} ({Q("ID_EPREUVE")}, {Q("ID_INSCRIPTION")}, {Q("VALEUR_NOTE")}, " +
            $"{Q("ABSENT")}, {Q("COPIE_NUM")}, {Q("DATE_SAISIE")}, {Q("CODE_UTR")}) VALUES (?, ?, ?, ?, ?, ?, ?)",
            P(note.IdEpreuve), P(note.IdInscription), P(note.ValeurNote), P(note.Absent ?? false),
            P(note.CopieNum), P(note.DateSaisie ?? DateTime.Now), P(note.CodeUtr));

    public int UpdateNoteExamen(NoteExamen note)
        => Execute(
            $"UPDATE {Q(Tables.NoteExamen)} SET {Q("VALEUR_NOTE")} = ?, {Q("ABSENT")} = ?, {Q("COPIE_NUM")} = ?, " +
            $"{Q("DATE_SAISIE")} = ?, {Q("CODE_UTR")} = ? WHERE {Q("ID_NOTE_EX")} = ?",
            P(note.ValeurNote), P(note.Absent ?? false), P(note.CopieNum),
            P(note.DateSaisie ?? DateTime.Now), P(note.CodeUtr), P(note.IdNoteEx));

    public int DeleteNoteExamen(int idNoteEx)
        => Execute($"DELETE FROM {Q(Tables.NoteExamen)} WHERE {Q("ID_NOTE_EX")} = ?", P(idNoteEx));

    /// <summary>Notes d'examen pondérables d'une classe pour une session (barème et coef inclus).</summary>
    public DataTable ListNotesPonderees(int idClasse, int idSession)
        => QueryTable(
            $"SELECT N.{Q("ID_INSCRIPTION")}, N.{Q("VALEUR_NOTE")}, N.{Q("ABSENT")}, " +
            $"EP.{Q("BAREME")}, EP.{Q("COEFFICIENT")}, EP.{Q("CODE_MATIERE")} " +
            $"FROM {Q(Tables.NoteExamen)} AS N " +
            $"INNER JOIN {Q(Tables.Epreuve)} AS EP ON N.{Q("ID_EPREUVE")} = EP.{Q("ID_EPREUVE")} " +
            $"WHERE EP.{Q("ID_CLASSE")} = ? AND EP.{Q("ID_SESSION")} = ?",
            P(idClasse), P(idSession));

    // ----- Mappers -----

    public static SessionExam MapSession(DataRow row) => new()
    {
        IdSession = DataRowMapper.GetInt32(row, "ID_SESSION"),
        IdAnnee = DataRowMapper.GetInt32(row, "ID_ANNEE"),
        Libelle = DataRowMapper.GetString(row, "LIBELLE"),
        Nature = DataRowMapper.GetString(row, "NATURE"),
        DateDebut = DataRowMapper.GetDateTime(row, "DATE_DEBUT"),
        DateFin = DataRowMapper.GetDateTime(row, "DATE_FIN"),
        Cloturee = DataRowMapper.GetBoolean(row, "CLOTUREE")
    };

    public static Epreuve MapEpreuve(DataRow row) => new()
    {
        IdEpreuve = DataRowMapper.GetInt32(row, "ID_EPREUVE"),
        IdSession = DataRowMapper.GetInt32(row, "ID_SESSION"),
        IdClasse = DataRowMapper.GetInt32(row, "ID_CLASSE"),
        CodeMatiere = DataRowMapper.GetString(row, "CODE_MATIERE"),
        DateEpreuve = DataRowMapper.GetDateTime(row, "DATE_EPREUVE"),
        HeureDebut = DataRowMapper.GetString(row, "HEURE_DEBUT"),
        DureeMn = DataRowMapper.GetInt32(row, "DUREE_MN"),
        Coefficient = DataRowMapper.GetDouble(row, "COEFFICIENT"),
        Bareme = DataRowMapper.GetDouble(row, "BAREME"),
        IdSalle = DataRowMapper.GetInt32(row, "ID_SALLE"),
        Surveillant = DataRowMapper.GetString(row, "SURVEILLANT")
    };

    public static EpreuveDetail MapEpreuveDetail(DataRow row) => new()
    {
        IdEpreuve = DataRowMapper.GetInt32(row, "ID_EPREUVE"),
        IdSession = DataRowMapper.GetInt32(row, "ID_SESSION"),
        Session = DataRowMapper.GetString(row, "SESSION_LIB"),
        IdClasse = DataRowMapper.GetInt32(row, "ID_CLASSE"),
        Classe = DataRowMapper.GetString(row, "CLASSE_LIB"),
        CodeMatiere = DataRowMapper.GetString(row, "CODE_MATIERE"),
        Matiere = DataRowMapper.GetString(row, "MATIERE_LIB"),
        DateEpreuve = DataRowMapper.GetDateTime(row, "DATE_EPREUVE"),
        HeureDebut = DataRowMapper.GetString(row, "HEURE_DEBUT"),
        DureeMn = DataRowMapper.GetInt32(row, "DUREE_MN"),
        Coefficient = DataRowMapper.GetDouble(row, "COEFFICIENT"),
        Bareme = DataRowMapper.GetDouble(row, "BAREME"),
        IdSalle = DataRowMapper.GetInt32(row, "ID_SALLE"),
        Salle = DataRowMapper.GetString(row, "SALLE_LIB"),
        Surveillant = DataRowMapper.GetString(row, "SURVEILLANT")
    };

    public static NoteExamen MapNoteExamen(DataRow row) => new()
    {
        IdNoteEx = DataRowMapper.GetInt32(row, "ID_NOTE_EX"),
        IdEpreuve = DataRowMapper.GetInt32(row, "ID_EPREUVE"),
        IdInscription = DataRowMapper.GetInt32(row, "ID_INSCRIPTION"),
        ValeurNote = DataRowMapper.GetDouble(row, "VALEUR_NOTE"),
        Absent = DataRowMapper.GetBoolean(row, "ABSENT"),
        CopieNum = DataRowMapper.GetString(row, "COPIE_NUM"),
        DateSaisie = DataRowMapper.GetDateTime(row, "DATE_SAISIE"),
        CodeUtr = DataRowMapper.GetString(row, "CODE_UTR")
    };
}

File: ltadmin/repositories/exam_repository.py
"""Dépôt des examens : sessions, épreuves, notes d'examen."""

from __future__ import annotations

import datetime as _dt
from typing import List, Optional

from ltadmin.data import row_mapper as rm
from ltadmin.data.schema import Tables
from ltadmin.models.dto import EpreuveDetail, NoteSaisieRow
from ltadmin.models.entities import Epreuve, NoteExamen, SessionExam
from ltadmin.repositories.base import RepositoryBase, p, q
from ltadmin.repositories.evaluation_repository import EvaluationRepository


class ExamRepository(RepositoryBase):

    # ----- SESSION_EXAM -----

    def get_session(self, id_session: int) -> Optional[SessionExam]:
        return self.query_single(
            f"SELECT * FROM {q(Tables.SESSION_EXAM)} WHERE {q('ID_SESSION')} = ?",
            ExamRepository.map_session, [p(id_session)])

    def list_sessions(self, id_annee: Optional[int] = None) -> List[SessionExam]:
        if id_annee is not None:
            return self.query_list(
                f"SELECT * FROM {q(Tables.SESSION_EXAM)} WHERE {q('ID_ANNEE')} = ? "
                f"ORDER BY {q('DATE_DEBUT')}, {q('LIBELLE')}",
                ExamRepository.map_session, [p(id_annee)])
        return self.query_list(
            f"SELECT * FROM {q(Tables.SESSION_EXAM)} ORDER BY {q('ID_ANNEE')} DESC, "
            f"{q('DATE_DEBUT')}",
            ExamRepository.map_session)

    def insert_session(self, session: SessionExam) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.SESSION_EXAM)} ({q('ID_ANNEE')}, {q('LIBELLE')}, "
            f"{q('NATURE')}, {q('DATE_DEBUT')}, {q('DATE_FIN')}, {q('CLOTUREE')}) "
            f"VALUES (?, ?, ?, ?, ?, ?)",
            [p(session.id_annee), p(session.libelle), p(session.nature),
             p(session.date_debut), p(session.date_fin),
             p(False if session.cloturee is None else session.cloturee)])

    def update_session(self, session: SessionExam) -> int:
        return self.execute(
            f"UPDATE {q(Tables.SESSION_EXAM)} SET {q('ID_ANNEE')} = ?, {q('LIBELLE')} = ?, "
            f"{q('NATURE')} = ?, {q('DATE_DEBUT')} = ?, {q('DATE_FIN')} = ?, {q('CLOTUREE')} = ? "
            f"WHERE {q('ID_SESSION')} = ?",
            [p(session.id_annee), p(session.libelle), p(session.nature),
             p(session.date_debut), p(session.date_fin),
             p(False if session.cloturee is None else session.cloturee), p(session.id_session)])

    def set_session_cloturee(self, id_session: int, cloturee: bool) -> int:
        return self.execute(
            f"UPDATE {q(Tables.SESSION_EXAM)} SET {q('CLOTUREE')} = ? WHERE {q('ID_SESSION')} = ?",
            [p(cloturee), p(id_session)])

    def delete_session(self, id_session: int) -> int:
        return self.execute(
            f"DELETE FROM {q(Tables.SESSION_EXAM)} WHERE {q('ID_SESSION')} = ?", [p(id_session)])

    # ----- EPREUVE -----

    def get_epreuve(self, id_epreuve: int) -> Optional[Epreuve]:
        return self.query_single(
            f"SELECT * FROM {q(Tables.EPREUVE)} WHERE {q('ID_EPREUVE')} = ?",
            ExamRepository.map_epreuve, [p(id_epreuve)])

    def get_epreuve_detail(self, id_epreuve: int) -> Optional[EpreuveDetail]:
        return self.query_single(
            ExamRepository.epreuve_detail_sql() + f" WHERE EP.{q('ID_EPREUVE')} = ?",
            ExamRepository.map_epreuve_detail, [p(id_epreuve)])

    def list_epreuves(self, id_session: Optional[int] = None,
                      id_classe: Optional[int] = None) -> List[EpreuveDetail]:
        conditions = []
        parameters = []
        if id_session is not None:
            conditions.append(f"EP.{q('ID_SESSION')} = ?")
            parameters.append(p(id_session))
        if id_classe is not None:
            conditions.append(f"EP.{q('ID_CLASSE')} = ?")
            parameters.append(p(id_classe))
        sql = ExamRepository.epreuve_detail_sql()
        if conditions:
            sql += " WHERE " + " AND ".join(conditions)
        sql += f" ORDER BY EP.{q('DATE_EPREUVE')}, EP.{q('ID_EPREUVE')}"
        return self.query_list(sql, ExamRepository.map_epreuve_detail, parameters)

    @staticmethod
    def epreuve_detail_sql() -> str:
        return (
            f"SELECT EP.*, SE.{q('LIBELLE')} AS SESSION_LIB, C.{q('LIBELLE')} AS CLASSE_LIB, "
            f"M.{q('MATIERE')} AS MATIERE_LIB, S.{q('NOM_SALLE')} AS SALLE_LIB "
            f"FROM (((({q(Tables.EPREUVE)} AS EP "
            f"INNER JOIN {q(Tables.SESSION_EXAM)} AS SE ON EP.{q('ID_SESSION')} = SE.{q('ID_SESSION')}) "
            f"INNER JOIN {q(Tables.CLASSE)} AS C ON EP.{q('ID_CLASSE')} = C.{q('ID_CLASSE')}) "
            f"INNER JOIN {q(Tables.MATIERE)} AS M ON EP.{q('CODE_MATIERE')} = M.{q('CODE_MATIERE')}) "
            f"LEFT JOIN {q(Tables.SALLE)} AS S ON EP.{q('ID_SALLE')} = S.{q('ID_SALLE')})"
        )

    def insert_epreuve(self, epreuve: Epreuve) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.EPREUVE)} ({q('ID_SESSION')}, {q('ID_CLASSE')}, "
            f"{q('CODE_MATIERE')}, {q('DATE_EPREUVE')}, {q('HEURE_DEBUT')}, {q('DUREE_MN')}, "
            f"{q('COEFFICIENT')}, {q('BAREME')}, {q('ID_SALLE')}, {q('SURVEILLANT')}) "
            f"VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            [p(epreuve.id_session), p(epreuve.id_classe), p(epreuve.code_matiere),
             p(epreuve.date_epreuve), p(epreuve.heure_debut), p(epreuve.duree_mn),
             p(epreuve.coefficient), p(epreuve.bareme), p(epreuve.id_salle),
             p(epreuve.surveillant)])

    def update_epreuve(self, epreuve: Epreuve) -> int:
        return self.execute(
            f"UPDATE {q(Tables.EPREUVE)} SET {q('ID_SESSION')} = ?, {q('ID_CLASSE')} = ?, "
            f"{q('CODE_MATIERE')} = ?, {q('DATE_EPREUVE')} = ?, {q('HEURE_DEBUT')} = ?, "
            f"{q('DUREE_MN')} = ?, {q('COEFFICIENT')} = ?, {q('BAREME')} = ?, {q('ID_SALLE')} = ?, "
            f"{q('SURVEILLANT')} = ? WHERE {q('ID_EPREUVE')} = ?",
            [p(epreuve.id_session), p(epreuve.id_classe), p(epreuve.code_matiere),
             p(epreuve.date_epreuve), p(epreuve.heure_debut), p(epreuve.duree_mn),
             p(epreuve.coefficient), p(epreuve.bareme), p(epreuve.id_salle),
             p(epreuve.surveillant), p(epreuve.id_epreuve)])

    def delete_epreuve(self, id_epreuve: int) -> int:
        return self.execute(
            f"DELETE FROM {q(Tables.EPREUVE)} WHERE {q('ID_EPREUVE')} = ?", [p(id_epreuve)])

    # ----- NOTE_EXAMEN -----

    def get_note_examen(self, id_epreuve: int, id_inscription: int) -> Optional[NoteExamen]:
        return self.query_single(
            f"SELECT * FROM {q(Tables.NOTE_EXAMEN)} WHERE {q('ID_EPREUVE')} = ? "
            f"AND {q('ID_INSCRIPTION')} = ?",
            ExamRepository.map_note_examen, [p(id_epreuve), p(id_inscription)])

    def list_grille_saisie(self, id_epreuve: int, id_classe: int) -> List[NoteSaisieRow]:
        return self.query_list(
            f"SELECT N.{q('ID_NOTE_EX')} AS ID_NOTE, I.{q('ID_INSCRIPTION')}, "
            f"E.{q('MATRICULE')}, E.{q('NOM')}, E.{q('PRENOM')}, N.{q('VALEUR_NOTE')}, "
            f"N.{q('ABSENT')} "
            f"FROM (({q(Tables.INSCRIPTION)} AS I "
            f"INNER JOIN {q(Tables.ETUDIANT)} AS E ON I.{q('ID_ETUDIANT')} = E.{q('ID_ETUDIANT')}) "
            f"LEFT JOIN {q(Tables.NOTE_EXAMEN)} AS N ON N.{q('ID_INSCRIPTION')} = "
            f"I.{q('ID_INSCRIPTION')} AND N.{q('ID_EPREUVE')} = ?) "
            f"WHERE I.{q('ID_CLASSE')} = ? ORDER BY E.{q('NOM')}, E.{q('PRENOM')}",
            EvaluationRepository.map_note_saisie_row, [p(id_epreuve), p(id_classe)])

    def count_notes_by_epreuve(self, id_epreuve: int) -> int:
        return self.scalar_int(
            f"SELECT COUNT(*) FROM {q(Tables.NOTE_EXAMEN)} WHERE {q('ID_EPREUVE')} = ?",
            [p(id_epreuve)])

    def insert_note_examen(self, note: NoteExamen) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.NOTE_EXAMEN)} ({q('ID_EPREUVE')}, {q('ID_INSCRIPTION')}, "
            f"{q('VALEUR_NOTE')}, {q('ABSENT')}, {q('COPIE_NUM')}, {q('DATE_SAISIE')}, "
            f"{q('CODE_UTR')}) VALUES (?, ?, ?, ?, ?, ?, ?)",
            [p(note.id_epreuve), p(note.id_inscription), p(note.valeur_note),
             p(False if note.absent is None else note.absent), p(note.copie_num),
             p(note.date_saisie or _dt.datetime.now()), p(note.code_utr)])

    def update_note_examen(self, note: NoteExamen) -> int:
        return self.execute(
            f"UPDATE {q(Tables.NOTE_EXAMEN)} SET {q('VALEUR_NOTE')} = ?, {q('ABSENT')} = ?, "
            f"{q('COPIE_NUM')} = ?, {q('DATE_SAISIE')} = ?, {q('CODE_UTR')} = ? "
            f"WHERE {q('ID_NOTE_EX')} = ?",
            [p(note.valeur_note), p(False if note.absent is None else note.absent),
             p(note.copie_num), p(note.date_saisie or _dt.datetime.now()),
             p(note.code_utr), p(note.id_note_ex)])

    def delete_note_examen(self, id_note_ex: int) -> int:
        return self.execute(
            f"DELETE FROM {q(Tables.NOTE_EXAMEN)} WHERE {q('ID_NOTE_EX')} = ?", [p(id_note_ex)])

    def list_notes_ponderes(self, id_classe: int, id_session: int) -> List[dict]:
        """Notes d'examen pondérables d'une classe pour une session."""
        return self.query_table(
            f"SELECT N.{q('ID_INSCRIPTION')}, N.{q('VALEUR_NOTE')}, N.{q('ABSENT')}, "
            f"EP.{q('BAREME')}, EP.{q('COEFFICIENT')}, EP.{q('CODE_MATIERE')} "
            f"FROM {q(Tables.NOTE_EXAMEN)} AS N "
            f"INNER JOIN {q(Tables.EPREUVE)} AS EP ON N.{q('ID_EPREUVE')} = EP.{q('ID_EPREUVE')} "
            f"WHERE EP.{q('ID_CLASSE')} = ? AND EP.{q('ID_SESSION')} = ?",
            [p(id_classe), p(id_session)])

    # ----- Mappers -----

    @staticmethod
    def map_session(row: dict) -> SessionExam:
        return SessionExam(
            id_session=rm.get_int(row, "ID_SESSION"),
            id_annee=rm.get_int(row, "ID_ANNEE"),
            libelle=rm.get_string(row, "LIBELLE"),
            nature=rm.get_string(row, "NATURE"),
            date_debut=rm.get_datetime(row, "DATE_DEBUT"),
            date_fin=rm.get_datetime(row, "DATE_FIN"),
            cloturee=rm.get_bool(row, "CLOTUREE"),
        )

    @staticmethod
    def map_epreuve(row: dict) -> Epreuve:
        return Epreuve(
            id_epreuve=rm.get_int(row, "ID_EPREUVE"),
            id_session=rm.get_int(row, "ID_SESSION"),
            id_classe=rm.get_int(row, "ID_CLASSE"),
            code_matiere=rm.get_string(row, "CODE_MATIERE"),
            date_epreuve=rm.get_datetime(row, "DATE_EPREUVE"),
            heure_debut=rm.get_string(row, "HEURE_DEBUT"),
            duree_mn=rm.get_int(row, "DUREE_MN"),
            coefficient=rm.get_double(row, "COEFFICIENT"),
            bareme=rm.get_double(row, "BAREME"),
            id_salle=rm.get_int(row, "ID_SALLE"),
            surveillant=rm.get_string(row, "SURVEILLANT"),
        )

    @staticmethod
    def map_epreuve_detail(row: dict) -> EpreuveDetail:
        return EpreuveDetail(
            id_epreuve=rm.get_int(row, "ID_EPREUVE"),
            id_session=rm.get_int(row, "ID_SESSION"),
            session=rm.get_string(row, "SESSION_LIB"),
            id_classe=rm.get_int(row, "ID_CLASSE"),
            classe=rm.get_string(row, "CLASSE_LIB"),
            code_matiere=rm.get_string(row, "CODE_MATIERE"),
            matiere=rm.get_string(row, "MATIERE_LIB"),
            date_epreuve=rm.get_datetime(row, "DATE_EPREUVE"),
            heure_debut=rm.get_string(row, "HEURE_DEBUT"),
            duree_mn=rm.get_int(row, "DUREE_MN"),
            coefficient=rm.get_double(row, "COEFFICIENT"),
            bareme=rm.get_double(row, "BAREME"),
            id_salle=rm.get_int(row, "ID_SALLE"),
            salle=rm.get_string(row, "SALLE_LIB"),
            surveillant=rm.get_string(row, "SURVEILLANT"),
        )

    @staticmethod
    def map_note_examen(row: dict) -> NoteExamen:
        return NoteExamen(
            id_note_ex=rm.get_int(row, "ID_NOTE_EX"),
            id_epreuve=rm.get_int(row, "ID_EPREUVE"),
            id_inscription=rm.get_int(row, "ID_INSCRIPTION"),
            valeur_note=rm.get_double(row, "VALEUR_NOTE"),
            absent=rm.get_bool(row, "ABSENT"),
            copie_num=rm.get_string(row, "COPIE_NUM"),
            date_saisie=rm.get_datetime(row, "DATE_SAISIE"),
            code_utr=rm.get_string(row, "CODE_UTR"),
        )
